### ISSUE TITLE
Threadpools from concurrent-ruby with Raven Sentry

### ISSUE THREAD

**Issue by kaspergrubbe** [getsentry/sentry-ruby#961]
Hi,

As the README says the proposed async method isn't optimal because launching a thread per exception can create an infinite number of threads.

A solution to that could be to launch the exceptions into Redis and let Resque or Sidekiq deal with it, this also comes with some tradeoffs since Redis storage can come at a premium.

Has anyone tried hooking `raven-ruby` up with the `ThreadPoolExecutor` from concurrent-ruby? (or something like that)? 

My naive approach was something like this:

```ruby
require 'concurrent'

Rails.configuration.sentry_pool = Concurrent::ThreadPoolExecutor.new(
  min_threads: 1,
  max_threads: 2,
  max_queue: 25,
  fallback_policy: :discard,
)

Raven.configure do |config|
  # (...)
  config.async = lambda do |event|
    Rails.configuration.sentry_pool.post do
      Raven.send_event(event)
    end
  end
end
```

But if this is so easy, I wonder why this gem isn't doing it already or why someone else aren't talking about such an approach. Do you have any feedback or insights to this?

Thanks,

KG.

**Comment by jeremy** [comment]
Nice tip KG. Does the executor drain automatically when the process terminates?

**Comment by kaspergrubbe** [comment]
@jeremy There are different operating modes, you can see the documentation here: https://ruby-concurrency.github.io/concurrent-ruby/1.1.6/Concurrent/ThreadPoolExecutor.html

You can shut down the threadpool with `#wait_for_termination` and it will drain and wait for the threadpool to complete.

However, I am not sure if such a hook exists in Unicorn (that we use).

**Comment by st0012** [comment]
hey @kaspergrubbe thanks for the suggestion!

> I wonder why this gem isn't doing it already or why someone else aren't talking about such an approach. 

I think it's partly because this gem was not actively developed in the past 2 years. But I'll definitely give it a try after cleaning up most of the issues 😉

**Comment by st0012** [comment]
@kaspergrubbe sorry for the long waiting time, but I managed to implement something similar in https://github.com/getsentry/sentry-ruby/pull/1155! it has been shipped in version `4.1.0`, please give it a try and let me know how it works for you 😉
